pkg/runtime/runtime.go

```diff
@@ -203,42 +193,64 @@ func (r *runtime) runCommand(ctx context.Context, cmd Command, addons []*addon.A
 		}
 		// TODO(dmitry-ilyevskiy): Print "live" status.
 		fmt.Printf("Configured addons:\n\t%s\n", strings.Join(lstMsgs, "\n\t"))
-	case InstallCommand:
 
+	case InstallCommand:
 		installAddonFn := func(a *addon.Addon) (err error) {
-			if !r.noSpin {
-				doneCh := make(chan *runStatus)
-
-				// Redirect stderr while we're spinning and filter out "proto: X"
-				// log spam (restore stderr on exit).
-				oldErr := os.Stderr
-				r, w, err := os.Pipe()
-				if err != nil {
-					return fmt.Errorf("can't create pipe2 for stderr redirect: %v", err)
+			pipeReader, pipeWriter := io.Pipe()
+			golog.SetOutput(pipeWriter)
+			buf := make([]byte, 1<<15)
+			var bufStr string
+
+			// Remove log spam from third-party libs.
+			appendFilterBufStr := func(n int) {
+				bufStr += string(buf[:n])
+				lines := strings.Split(bufStr, "\n")
+				if len(lines) == 1 {
+					return
 				}
-				os.Stderr = w
-
-				defer func() {
-					// Restor stderr and close pipe.
-					os.Stderr = oldErr
-					w.Close()
-
-					sc := bufio.NewScanner(r)
-					s := &runStatus{err: err, doneCh: make(chan struct{})}
-					for sc.Scan() {
-						if strings.Contains(sc.Text(), "proto: ") {
-							continue
+				// Prints all lines but the last line.
+				for i := 0; i < len(lines)-1; i++ {
+					line := lines[i]
+					if strings.Contains(line, "proto: ") || line == "" {
+						continue
+					}
+					fmt.Println(line)
+				}
+				// Since the last line might be incomplete, keep it until
+				// we see a newline.
+				bufStr = lines[len(lines)-1]
+			}
+			go func() {
+				for {
+					n, err := pipeReader.Read(buf)
+					if err != nil {
+						if err == io.EOF {
+							appendFilterBufStr(n)
+							break
 						}
-						s.msgs = append(s.msgs, sc.Text())
+						log.Errorf("error while reading from filtering pipe: %v", err)
+						break
 					}
+					appendFilterBufStr(n)
+					time.Sleep(100 * time.Millisecond)
+				}
+			}()
 
-					doneCh <- s
-					<-s.doneCh
-				}()
+			defer func() {
+				// Restore go log output and close pipe.
+				golog.SetOutput(os.Stderr)
+				pipeWriter.Close()
+			}()
 
-				go spinMsg(a.Name, doneCh)
+			if r.noSpin {
+				return a.Install(ctx)
 			}
-			return a.Install(ctx)
+
+			errCh := make(chan error)
+			go spinMsg(a.Name, errCh)
+			err = a.Install(ctx)
+			errCh <- err
```

Comment:
before sending this error, should we also wait until the pipeReader.Read(buf) reaches EOF?